File: src/project/onboarding.ts
import type { LarkChannel } from '@larksuiteoapi/node-sdk';
import { log } from '../core/logger';
import { buildWelcomeCard } from '../card/command-cards';
import { createBackend } from '../agent';
import { defaultNoMention, type Project } from './registry';

/** 项目后端的能力位（用于欢迎卡按后端裁剪命令）。未知/手编 id 解析失败 ⇒ undefined
 * （= 全列，与 codex 一致），绝不因此让建项目/绑群失败。 */
function welcomeCaps(backend?: string): { goal?: boolean; compact?: boolean; resume?: boolean } | undefined {
  try {
    return createBackend(backend).capabilities;
  } catch {
    return undefined;
  }
}

/**
 * Public command manual — an internet-readable Feishu doc (full command guide +
 * product intro). Linked from the welcome card and surfaced as a chat tab.
 * Empty string = skip the tab + the manual button (no broken links). Set this
 * to the doc's share URL once it's published as "互联网获取链接的人可阅读".
 */
const HELP_DOC_URL: string = 'https://my.feishu.cn/wiki/PZ23wGr7JiKK5RkIG4rcZXzGn5g';

/**
 * PC 端深链：让群菜单的链接在飞书**侧边栏**打开而不是弹独立浏览器（官方
 * `applink … mode=sidebar-semi` 前缀，见群菜单概述文档）。仅用于 menu_tree 的
 * `pc_url`；移动端走 `common_url` 原链接。Exported for tests.
 */
export function sidebarPcUrl(url: string): string {
  return `https://applink.feishu.cn/client/web_url/open?mode=sidebar-semi&url=${encodeURIComponent(url)}`;
}

/**
 * Onboard a freshly-created project group (design: 项目=群): post a welcome card
 * listing every command this group supports, Pin it (so it lives in the chat's
 * Pins tab), and add a "👈 查看可使用的命令" url tab pointing at the public manual.
 *
 * This is distinct from the 群公告 top banner ({@link setAnnouncement}): the
 * banner is the always-on one-line project header, while the Pin'd card + tab
 * are the command cheat-sheet. Every step is best-effort — the group is still
 * usable if any of them fails (e.g. missing scope).
 */
export async function onboardGroup(channel: LarkChannel, project: Project): Promise<void> {
  const kind = project.kind ?? 'multi';
  const chatId = project.chatId;
  // In a 'joined' group the bot is a plain member, not the owner: don't Pin or
  // add a chat tab (those mutate the group's structure and a member may lack the
  // permission anyway). The one-off welcome card is still sent — it's just a
  // message. Created groups keep the full treatment (welcome → Pin → tab).
  const decorate = (project.origin ?? 'created') !== 'joined';

  // 1. Welcome card. Raw interactive JSON (no callback buttons to mutate, just
  //    an open_url link), so it needn't be a CardKit entity. For created groups
  //    we capture the message_id and Pin it; joined groups skip the Pin.
  try {
    const noMention = project.noMention ?? defaultNoMention(project);
    const content = JSON.stringify(buildWelcomeCard(kind, HELP_DOC_URL || undefined, noMention, welcomeCaps(project.backend)));
    const sent = await channel.rawClient.im.v1.message.create({
      params: { receive_id_type: 'chat_id' },
      data: { receive_id: chatId, msg_type: 'interactive', content },
    });
    const messageId = (sent as { data?: { message_id?: string } }).data?.message_id;
    if (messageId && decorate) {
      await channel.rawClient.im.v1.pin.create({ data: { message_id: messageId } });
      log.info('project', 'onboard-pin', { name: project.name });
    }
  } catch (err) {
    log.fail('project', err, { phase: 'onboard-welcome' });
  }

  // 2. Chat tab → public manual. Skipped for joined groups (don't mutate the
  //    group) and when no URL is configured (otherwise we'd pin a tab to an
  //    empty link). Sits to the right of the built-in Pins tab. Only doc/url
  //    tab types are allowed by the API.
  if (decorate && HELP_DOC_URL) {
    try {
      await channel.rawClient.im.v1.chatTab.create({
        path: { chat_id: chatId },
        data: {
          chat_tabs: [{ tab_name: '👈 使用说明', tab_type: 'url', tab_content: { url: HELP_DOC_URL } }],
        },
      });
      log.info('project', 'onboard-tab', { name: project.name });
    } catch (err) {
      log.fail('project', err, { phase: 'onboard-tab' });
    }
  }

  // 3. 群菜单「🤖 Codex」→ 命令手册（M-6 群内可发现性）。群菜单常驻输入框上方，
  //    可见性远高于群 Tab；动作只支持 REDIRECT_LINK —— PC 端加 applink
  //    sidebar-semi 前缀在侧边栏打开，移动端走原链接。需可选 scope
  //    im:chat.menu_tree:write_only：未开通时 create 报错 → log + 跳过，
  //    其余 onboarding 不受影响（与 Pin/Tab 一致的 best-effort 语义）。
  if (decorate && HELP_DOC_URL) {
    try {
      await channel.rawClient.im.v1.chatMenuTree.create({
        path: { chat_id: chatId },
        data: {
          menu_tree: {
            chat_menu_top_levels: [
              {
                chat_menu_item: {
                  action_type: 'REDIRECT_LINK',
                  name: '🤖 Codex',
                  redirect_link: { common_url: HELP_DOC_URL, pc_url: sidebarPcUrl(HELP_DOC_URL) },
                },
              },
            ],
          },
        },
      });
      log.info('project', 'onboard-menu', { name: project.name });
    } catch (err) {
      log.fail('project', err, { phase: 'onboard-menu' });
    }
  }
}

File: internal/bot/onboarding.go
package bot

// onboarding.go —— 建群后的 onboarding 编排（对齐 TS project/onboarding.ts 的 onboardGroup）：
// 1) 发「🤖 本群使用说明」欢迎卡（card.BuildWelcomeCard，经 SendCardFunc 走 CardKit 实体路径）；
// 2) created 群额外 Pin 该卡 + 加群 Tab「👈 使用说明」+ 加群菜单「🤖 <后端名>」→ 手册。
// joined 群（bot 只是普通成员）只发欢迎卡，不 Pin/Tab/Menu（与 TS 一致）。
// 全部 best-effort：任何一步失败仅告警，不阻断建群。

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/modelzen/feishu-codex-bridge/internal/agent"
	"github.com/modelzen/feishu-codex-bridge/internal/card"
	"github.com/modelzen/feishu-codex-bridge/internal/core"
	"github.com/modelzen/feishu-codex-bridge/internal/project"
)

// helpDocURL 公开命令手册（互联网可读飞书文档）。对齐 TS onboarding.ts 的 HELP_DOC_URL。
const helpDocURL = "https://my.feishu.cn/wiki/PZ23wGr7JiKK5RkIG4rcZXzGn5g"

// onboardGroup 建群/绑群后的 onboarding（对齐 TS onboardGroup）。
func (o *Orchestrator) onboardGroup(ctx context.Context, p project.Project, chatID string) {
	if o.SendCardFunc == nil {
		core.Warn(ctx, "bot", "onboard", "SendCardFunc 未注入，欢迎卡无法发出")
		return
	}
	agentName := backendDisplayName(p.Backend)
	caps := welcomeCaps(p.Backend)
	noMention := effectiveNoMention(p)
	welcome := card.BuildWelcomeCard(p.Kind, helpDocURL, noMention, caps, agentName)
	jsonBytes, err := json.Marshal(welcome)
	if err != nil {
		core.Warn(ctx, "bot", "onboard", "欢迎卡序列化失败："+err.Error())
		return
	}
	msgID, err := o.SendCardFunc(ctx, chatID, jsonBytes)
	if err != nil {
		core.Warn(ctx, "bot", "onboard", "欢迎卡发送失败（可忽略）："+err.Error())
		return
	}
	core.Info(ctx, "bot", "onboard", "欢迎卡已发送 chatID="+chatID+" msgID="+msgID)

	// joined 群（bot 是普通成员）不 Pin/Tab/Menu —— 与 TS 一致。
	if !project.ShouldOnboard(p) {
		return
	}
	// 2a. Pin 欢迎卡。
	if err := o.pinMessage(ctx, msgID); err != nil {
		core.Warn(ctx, "bot", "onboard-pin", "欢迎卡 Pin 失败（可忽略，可能缺 im:chat:pin）："+err.Error())
	} else {
		core.Info(ctx, "bot", "onboard-pin", "欢迎卡已 Pin chatID="+chatID)
	}
	// 2b. 群 Tab「👈 使用说明」→ 手册。
	if err := o.addChatTab(ctx, chatID, "👈 使用说明", helpDocURL); err != nil {
		core.Warn(ctx, "bot", "onboard-tab", "群 Tab 失败（可忽略，可能缺 im:chat.tabs:write_only）："+err.Error())
	} else {
		core.Info(ctx, "bot", "onboard-tab", "群 Tab 已添加 chatID="+chatID)
	}
	// 2c. 群菜单「🤖 <后端名>」→ 手册（PC 端侧边栏打开）。
	if err := o.addChatMenu(ctx, chatID, "🤖 "+agentName, helpDocURL, project.SidebarPcUrl(helpDocURL)); err != nil {
		core.Warn(ctx, "bot", "onboard-menu", "群菜单失败（可忽略，可能缺 im:chat.menu_tree:write_only）："+err.Error())
	} else {
		core.Info(ctx, "bot", "onboard-menu", "群菜单已添加 chatID="+chatID)
	}
}

// pinMessage / addChatTab / addChatMenu 经局部接口断言调用 feishu.Channel 的原生方法。

func (o *Orchestrator) pinMessage(ctx context.Context, messageID string) error {
	type pinner interface{ PinMessage(context.Context, string) error }
	p, ok := o.Channel.(pinner)
	if !ok {
		return fmt.Errorf("Channel 未实现 PinMessage")
	}
	return p.PinMessage(ctx, messageID)
}

func (o *Orchestrator) addChatTab(ctx context.Context, chatID, name, url string) error {
	type tabber interface{ AddChatTab(context.Context, string, string, string) error }
	t, ok := o.Channel.(tabber)
	if !ok {
		return fmt.Errorf("Channel 未实现 AddChatTab")
	}
	return t.AddChatTab(ctx, chatID, name, url)
}

func (o *Orchestrator) addChatMenu(ctx context.Context, chatID, name, url, pcURL string) error {
	type menuer interface{ AddChatMenu(context.Context, string, string, string, string) error }
	m, ok := o.Channel.(menuer)
	if !ok {
		return fmt.Errorf("Channel 未实现 AddChatMenu")
	}
	return m.AddChatMenu(ctx, chatID, name, url, pcURL)
}

// backendDisplayName 本群后端的展示名（Codex / Claude …）；未知/未设 → Codex（向后兼容）。
func backendDisplayName(backend string) string {
	if backend == "" {
		backend = agent.DEFAULT_BACKEND_ID
	}
	if e, ok := agent.CatalogByID(backend); ok {
		return e.DisplayName
	}
	return "Codex"
}

// welcomeCaps 按后端能力裁剪欢迎卡命令（对齐 TS welcomeCaps）。
// 未知/手编 id 解析失败 → 返回零值 HelpCaps（nil 字段 ⇒ 全列，与 codex 一致）。
func welcomeCaps(backend string) card.HelpCaps {
	if backend == "" {
		backend = agent.DEFAULT_BACKEND_ID
	}
	b, err := agent.CreateBackend(backend)
	if err != nil {
		return card.HelpCaps{}
	}
	c := b.Capabilities()
	return card.HelpCaps{Goal: boolPtr(c.Goal), Compact: boolPtr(c.Compact), Resume: boolPtr(c.Resume)}
}

// effectiveNoMention 群有效免@状态（项目显式设置优先，否则取默认）。
func effectiveNoMention(p project.Project) bool {
	if p.NoMention != nil {
		return *p.NoMention
	}
	return project.DefaultNoMention(p)
}
